File: Java/MobileAzureDevDays/app/src/main/java/com/xamarin/azuredevdays/MainActivity.java
package com.xamarin.azuredevdays;

import android.app.DialogFragment;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.app.Dialog;
import android.app.AlertDialog;
import android.content.DialogInterface;
import java.util.HashMap;
import java.util.Map;

import com.google.gson.Gson;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;

public class MainActivity extends AppCompatActivity implements OnClickListener {

    SentimentClient sentimentClient;
    Gson gson = new Gson();

    Toolbar toolbar;
    LinearLayout backgroundLayout;
    EditText sentimentText;
    Button getSentimentButton;
    Button crashButton;
    Button eventButton;
    Button colorButton;
    TextView emojiView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        sentimentClient = new SentimentClient();
        try {
            sentimentClient.GetKey(
                    new Callback() {
                        @Override
                        public void onFailure(Call call, IOException e) {
                            System.out.println("Error getting key!");
                        }
                        @Override
                        public void onResponse(Call call, Response response) throws IOException {
                            if (response.isSuccessful()) {
                                Models.KeyResponse keyResponse = gson.fromJson(response.body().string(),Models.KeyResponse.class);
                                sentimentClient.sentimentAPIRegion = keyResponse.region;
                                sentimentClient.sentimentAPIKey = keyResponse.key;
                            }
                        }
                    }
            );
        } catch (IOException e) {
            e.printStackTrace();
        }

        int defaultBkgColor = Color.parseColor(Constants.Colors.defaultBackgroundColor);

        backgroundLayout = (LinearLayout) findViewById(R.id.backGroundLayout);
        backgroundLayout.setBackgroundColor(defaultBkgColor);
        toolbar.setBackgroundColor(defaultBkgColor);
        getWindow().setStatusBarColor(darkenColor(defaultBkgColor));

        sentimentText = (EditText) findViewById(R.id.sentimentText);
        emojiView = (TextView) findViewById(R.id.emojiView);

        getSentimentButton = (Button) findViewById(R.id.getSentimentButton);
        getSentimentButton.setOnClickListener(this);

        Button crashButton = (Button) findViewById(R.id.sendCrashButton);
        crashButton.setOnClickListener(this);

        Button eventButton = (Button) findViewById(R.id.sendEventButton);
        eventButton.setOnClickListener(this);

        Button colorButton = (Button) findViewById(R.id.sendColorButton);
        colorButton.setOnClickListener(this);


    }

    public void onClick(final View view){
        switch(view.getId()){
            case R.id.getSentimentButton:

                if(sentimentText.getText().toString().equals("")) {
                    Snackbar.make(view, "You must enter something!", Snackbar.LENGTH_SHORT).show();
                    return;
                }

                String res = "";
                getSentimentButton.setText("Calculating");
                getSentimentButton.setEnabled(false);
                sentimentText.setEnabled(false);
                try {
                    sentimentClient.GetSentimentResult(sentimentText.getText().toString(),
                            new Callback() {
                                @Override
                                public void onFailure(Call call, IOException e) {
                                    // Something went wrong
                                }

                                @Override
                                public void onResponse(Call call, Response response) throws IOException {

                                    if (response.isSuccessful()) {

                                        Models.SentimentResponse sentimentResponse = gson.fromJson(response.body().string(),Models.SentimentResponse.class);

                                        final float score = sentimentResponse.documents.get(0).score;

                                        MainActivity.this.runOnUiThread(new Runnable() {
                                            @Override
                                            public void run() {
                                                UpdateUI(score);
                                            }
                                        });

                                    } else {
                                        Snackbar.make(view, response.message(), Snackbar.LENGTH_LONG).show();
                                        MainActivity.this.runOnUiThread(new Runnable() {
                                            @Override
                                            public void run() {
                                                UpdateUI(-1);
                                            }
                                        });
                                    }
                                }
                            });
                } catch (IOException e) {
                    e.printStackTrace();
                }

                break;
            case R.id.sendCrashButton:
                DialogFragment crashDialog = new CrashDialog();
                crashDialog.show(getFragmentManager(), "crashDialog");
                break;

            case R.id.sendEventButton:
                DialogFragment eventDialog = new EventDialog();
                eventDialog.show(getFragmentManager(), "eventDialog");
                break;

            case R.id.sendColorButton:
                DialogFragment colorDialog = new ColorDialog();
                colorDialog.show(getFragmentManager(), "colorDialog");
                break;
        }
    }


    public void UpdateUI(float score){

        if(score!=-1) {
            int sentimentColor = Color.parseColor(getBackgroundColor(score));
            emojiView.setText(getEmojiString(score));
            backgroundLayout.setBackgroundColor(sentimentColor);
            toolbar.setBackgroundColor(sentimentColor);
            MainActivity.this.getWindow().setStatusBarColor(darkenColor(sentimentColor));
        }
        sentimentText.setEnabled(true);
        getSentimentButton.setText("Submit");
        getSentimentButton.setEnabled(true);
    }

    public String getEmojiString(float score){

        if(score < 0.4)
        {
            return Constants.Emojis.SadFaceEmoji;
        }
        else if (score >= 0.4 && score <=0.6)
        {
            return Constants.Emojis.NeutralFaceEmoji;
        }
        else if (score > 0.6){
            return Constants.Emojis.HappyFaceEMoji;
        }
        else{
            return "";
        }
    }

    public String getBackgroundColor(float score){
        if(score <= 0.1)
            return Constants.Colors.EmotionColor1;
        else if (score > 0.1 && score <= 0.2)
            return Constants.Colors.EmotionColor2;
        else if (score > 0.2 && score <= 0.3)
            return Constants.Colors.EmotionColor3;
        else if (score > 0.3 && score <= 0.4)
            return Constants.Colors.EmotionColor4;
        else if (score > 0.4 && score <= 0.6)
            return Constants.Colors.EmotionColor5;
        else if (score > 0.6 && score <= 0.7)
            return Constants.Colors.EmotionColor6;
        else if (score > 0.7 && score <= 0.8)
            return Constants.Colors.EmotionColor7;
        else if (score > 0.8 && score <= 0.9)
            return Constants.Colors.EmotionColor8;
        else if (score > 0.9)
            return Constants.Colors.EmotionColor9;
        else
            return "";
    }

    private static int darkenColor(int color) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[2] *= 0.8f;
        return Color.HSVToColor(hsv);
    }

    public static class CrashDialog extends DialogFragment {
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setMessage("A crash report will be sent when you reopen the app.")
                    .setPositiveButton("Crash app", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int id) {
                            throw new RuntimeException("crashing");
                        }
                    }).setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                }
            });
            return builder.create();
        }
    }

    public static class EventDialog extends DialogFragment {
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setMessage("Event sent").setPositiveButton("OK", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                }
            });
            return builder.create();
        }
    }

    private static Map<String, String> properties = new HashMap<>();

    public static class ColorDialog extends DialogFragment {
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            CharSequence[] colors = {"Yellow", "Blue", "Red"};
            builder.setTitle("Pick a color").setItems(colors, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int index) {
                    switch (index) {
                        case 0:
                            properties.put("Color", "Yellow");
                            break;
                        case 1:
                            properties.put("Color", "Blue");
                            break;
                        case 2:
                            properties.put("Color", "Red");
                            break;
                    }
                }
            });
            return builder.create();
        }
    }

}

File: Java/MobileAzureDevDays/app/src/main/java/com/xamarin/azuredevdays/MainActivity.kt
package com.xamarin.azuredevdays

import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.View.OnClickListener

import com.google.gson.Gson
import com.xamarin.azuredevdays.model.KeyResponse
import com.xamarin.azuredevdays.model.SentimentResponse

import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*

import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response

import java.io.IOException
import java.util.HashMap

private val PROPERTIES = HashMap<String, String>()

class MainActivity : AppCompatActivity(), OnClickListener {

    private val sentimentClient = SentimentClient()
    private val gson = Gson()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        try {
            sentimentClient.getKey(
                object : Callback {
                    override fun onFailure(call: Call, e: IOException) {
                        println("Error getting key!")
                    }

                    @Throws(IOException::class)
                    override fun onResponse(call: Call, response: Response) {
                        if (response.isSuccessful) response.body()?.let { responseBody ->
                            gson.fromJson(responseBody.string(), KeyResponse::class.java)?.let {
                                sentimentClient.sentimentApiRegion = it.region ?: ""
                                sentimentClient.sentimentApiKey = it.key ?: ""
                            }
                        }
                    }
                }
            )
        } catch (e: IOException) {
            e.printStackTrace()
        }

        val defaultBkgColor = Color.parseColor(DEFAULT_BACKGROUND_COLOR)
        backgroundLayout?.setBackgroundColor(defaultBkgColor)
        toolbar?.setBackgroundColor(defaultBkgColor)
        window.statusBarColor = darkenColor(defaultBkgColor)

        getSentimentButton?.setOnClickListener(this)
        sendCrashButton?.setOnClickListener(this)
        sendEventButton?.setOnClickListener(this)
        sendColorButton?.setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.getSentimentButton -> {
                if (sentimentText?.text?.isBlank() == true) {
                    Snackbar.make(view, "You must enter something!", Snackbar.LENGTH_SHORT).show()
                    return
                }

                getSentimentButton?.text = "Calculating"
                getSentimentButton?.isEnabled = false
                sentimentText?.isEnabled = false

                try {
                    sentimentClient.getSentimentResult(sentimentText?.text.toString(),
                        object : Callback {
                            override fun onFailure(call: Call, e: IOException) {
                                // Something went wrong
                            }

                            @Throws(IOException::class)
                            override fun onResponse(call: Call, response: Response) {

                                if (response.isSuccessful) response.body()?.let { responseBody ->
                                    val sentimentResponse: SentimentResponse = gson.fromJson(
                                        responseBody.string(),
                                        SentimentResponse::class.java
                                    )
                                    val score = sentimentResponse.documents[0].score
                                    this@MainActivity.runOnUiThread { updateUI(score) }
                                } else {
                                    Snackbar.make(view, response.message(), Snackbar.LENGTH_LONG)
                                        .show()
                                    this@MainActivity.runOnUiThread { updateUI(-1f) }
                                }
                            }
                        })
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
            R.id.sendCrashButton -> CrashDialog().show(fragmentManager, "crashDialog")
            R.id.sendEventButton -> EventDialog().show(fragmentManager, "eventDialog")
            R.id.sendColorButton -> ColorDialog().show(fragmentManager, "colorDialog")
        }
    }

    fun updateUI(score: Float) {
        if (score != -1f) {
            val sentimentColor = Color.parseColor(getBackgroundColor(score))
            emojiView?.text = getEmojiString(score)
            backgroundLayout?.setBackgroundColor(sentimentColor)
            toolbar?.setBackgroundColor(sentimentColor)
            this@MainActivity.window.statusBarColor = darkenColor(sentimentColor)
        }

        sentimentText?.isEnabled = true
        getSentimentButton?.text = "Submit"
        getSentimentButton?.isEnabled = true
    }

    private fun getEmojiString(score: Float) = when {
        score < 0.4 -> SAD_FACE_EMOJI
        score in 0.4..0.6 -> NEUTRAL_FACE_EMOJI
        score > 0.6 -> HAPPY_FACE_EMOJI
        else -> ""
    }

    private fun getBackgroundColor(score: Float) = when {
        score <= 0.1 -> EMOTION_COLOR_1
        score > 0.1 && score <= 0.2 -> EMOTION_COLOR_2
        score > 0.2 && score <= 0.3 -> EMOTION_COLOR_3
        score > 0.3 && score <= 0.4 -> EMOTION_COLOR_4
        score > 0.4 && score <= 0.6 -> EMOTION_COLOR_5
        score > 0.6 && score <= 0.7 -> EMOTION_COLOR_6
        score > 0.7 && score <= 0.8 -> EMOTION_COLOR_7
        score > 0.8 && score <= 0.9 -> EMOTION_COLOR_8
        score > 0.9 -> EMOTION_COLOR_9
        else -> ""
    }

    private fun darkenColor(color: Int): Int {
        val hsv = FloatArray(3)
        Color.colorToHSV(color, hsv)
        hsv[2] *= 0.8f
        return Color.HSVToColor(hsv)
    }

    class CrashDialog : DialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog =
            AlertDialog.Builder(activity).apply {
                setMessage("A crash report will be sent when you reopen the app.")
                setPositiveButton("Crash app") { _, _ -> throw RuntimeException("crashing") }
                setNegativeButton("Cancel") { _, _ -> }
            }.create()
    }

    class EventDialog : DialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog =
            AlertDialog.Builder(activity).apply {
                setMessage("Event sent").setPositiveButton("OK") { _, _ -> }
            }.create()
    }

    class ColorDialog : DialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog =
            AlertDialog.Builder(activity).apply {
                val colors = arrayOf<CharSequence>("Yellow", "Blue", "Red")
                setTitle("Pick a color").setItems(colors) { _, index ->
                    when (index) {
                        0 -> PROPERTIES["Color"] = "Yellow"
                        1 -> PROPERTIES["Color"] = "Blue"
                        2 -> PROPERTIES["Color"] = "Red"
                    }
                }
            }.create()
    }
}
